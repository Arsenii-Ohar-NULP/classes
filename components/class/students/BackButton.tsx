import React from 'react';
import {useRouter} from "next/router";
import {Button} from "react-bootstrap";
import backIcon from 'icons/back.svg';
import Image from "next/image";

export const BackButton = () => {
    const router = useRouter();

    const goBack = async () => {
        router.back();
    }

    return <Button variant={'outline-light'} className={'p-1 rounded-pill'} onClick={goBack}>
        <Image src={backIcon} alt={'Back'} height={24} width={24} />
    </Button>
}