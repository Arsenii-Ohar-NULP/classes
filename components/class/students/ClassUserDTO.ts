export interface ClassUserDTO{
        class: number;
        student: number;
}

