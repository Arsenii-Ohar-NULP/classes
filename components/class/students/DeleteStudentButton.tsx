import React from 'react';
import {Button} from "react-bootstrap";
import deleteIcon from 'icons/delete.svg';
import Image from "next/image";

interface DeleteStudentButtonProps {
    onDelete: VoidFunction
}

export const DeleteStudentButton = ({onDelete}: DeleteStudentButtonProps) => {
    return <Button onClick={onDelete} variant={'danger'} className={'p-1'}>
        <Image src={deleteIcon} alt={'Delete Icon'} height={24} width={24}/>
    </Button>
}