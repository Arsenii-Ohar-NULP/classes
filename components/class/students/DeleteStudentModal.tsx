import React, {useState} from "react";
import {useRouter} from "next/navigation";
import {useAppDispatch} from "components/redux/store";
import InvalidCredentials from "components/errors/InvalidCredentials";
import {logout} from "components/login/AuthService";
import {Button, Modal} from "react-bootstrap";
import {ClassUserDTO, removeStudentFromClass} from "components/class/students/StudentsService";

export default function DeleteStudentModal({
  classUser,
  onDelete,
  show,
  close,
}: {
  classUser: ClassUserDTO;
  onDelete: () => void;
  show: boolean;
  close: () => void;
}) {
  const [isDeleting, setIsDeleting] = useState<boolean>(false);

  const router = useRouter();
  const dispatch = useAppDispatch();

  function deleteStudent() {
    setIsDeleting(true);
    removeStudentFromClass(classUser)
      .then(() => {
        onDelete();
        close();
      })
      .catch((error) => {
        if (error instanceof InvalidCredentials) {
          logout(dispatch, router);
        }
      })
      .finally(() => setIsDeleting(false));
  }

  return (
    <>
      <Modal show={show} backdrop={true} onHide={close} keyboard={false}>
        <Modal.Header closeButton={!isDeleting}>
          <Modal.Title>Are you sure?</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          The student will be deleted and there is no way to undo this. Are you
          sure?
        </Modal.Body>
        <Modal.Footer>
          <Button
            variant="danger"
            onClick={deleteStudent}
            disabled={isDeleting}
          >
            Delete
          </Button>
          <Button variant="success" onClick={close} disabled={isDeleting}>
            No, take me back
          </Button>
        </Modal.Footer>
      </Modal>
    </>
  );
}
