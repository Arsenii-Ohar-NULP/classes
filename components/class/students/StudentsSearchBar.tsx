import React, {ChangeEvent, useEffect} from 'react';
import {FormControl} from "react-bootstrap";
import {useAppDispatch} from "../../redux/store";
import {searchActions} from "../../redux/search";


export const StudentsSearchBar = () => {
    const dispatch = useAppDispatch();

    useEffect(() => {
        dispatch(searchActions.searchStudents(null));

        return () => {
            dispatch(searchActions.searchStudents(null));
        }
    }, [])

    const onFieldChange = (e: ChangeEvent<HTMLInputElement>) => {
        dispatch(searchActions.searchStudents(e.target.value));
    }

    return <FormControl className={'mb-3'}
                        type={'text'}
                        onChange={onFieldChange}
                        placeholder={'Search for a student'}
    />

}