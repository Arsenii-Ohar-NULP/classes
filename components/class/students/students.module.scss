/* Hide scrollbar for Chrome, Safari and Opera */
.hide-scrollbar::-webkit-scrollbar {
  display: none;
}

/* Hide scrollbar for IE, Edge and Firefox */
.hide-scrollbar {
  -ms-overflow-style: none;  /* IE and Edge */
  scrollbar-width: none;  /* Firefox */
}

.smooth-shadow{
  box-shadow: rgba(47, 43, 61, 0.1) 0px 4px 18px 0px;
}

.neat-text {
  color: rgba(47, 43, 61, 0.78);
}

.student{}

.student:hover{
  background-color: #eafdd8;
}