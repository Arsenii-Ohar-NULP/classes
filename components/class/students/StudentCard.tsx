import React, {useState} from 'react';
import User from "../../account/User";
import clsx from "clsx";
import ProfilePicture from "../../ProfilePicture";
import styles from 'components/class/students/students.module.scss';
import {DeleteStudentButton} from "./DeleteStudentButton";
import {VisibleOnHover} from "./VisibleOnHover";

interface StudentCardProps {
    student: User;
    onDelete: VoidFunction
}

export const StudentCard = ({student, onDelete}: StudentCardProps) => {
    const [isHover, setIsHover] = useState<boolean>();
    return <tr
        className={clsx('p-2 w-100 text-start overflow-scroll', styles['neat-text'], styles['student'])}
        onMouseOver={() => setIsHover(true)}
        onMouseOut={() => setIsHover(false)}>
        <td scope={'row'} className={'d-flex text-center gap-2 align-items-center'}>
            <ProfilePicture user={student} hoverOn={false}/>
            <div>{student.firstName} {student.lastName}</div>
        </td>
        <td scope={'row'}>{student.email}</td>
        <td scope={'row'}>
            <div className={'d-flex gap-3'}>
                {student.phone}
                <VisibleOnHover isHover={isHover}>
                    <DeleteStudentButton onDelete={onDelete}/>
                </VisibleOnHover>
            </div>
        </td>
    </tr>
}