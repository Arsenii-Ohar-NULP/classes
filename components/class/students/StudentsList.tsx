import React, {useState} from 'react';
import User from "components/account/User";
import {StudentCard} from "components/class/students/StudentCard";
import {StudentsListHeader} from "components/class/students/StudentsListHeader";
import clsx from "clsx";
import styles from 'components/class/students/students.module.scss';
import {useAppSelector} from "../../redux/store";
import Class from "components/classes/Class";
import DeleteStudentModal from "./DeleteStudentModal";

interface StudentsListProps {
    students: User[];
    isLoading: boolean;
    cls: number;
    onDelete: (student: number) => void;
}

export const StudentsList = ({students, isLoading, cls, onDelete}: StudentsListProps) => {
    const [deleteStudentId, setDeleteStudentId] = useState<number>(-1);

    return (<div className={'container'}>
        {!isLoading && (students == null || students.length == 0) ?
            <p className={'text-center'}>There are no students in this class.</p> : !isLoading &&
            <div className={clsx('overflow-scroll', styles['hide-scrollbar'], styles['smooth-shadow'])}>
                <table className={'table shadow m-0'}>
                    <StudentsListHeader/>
                    <tbody>
                    {students?.map((student) => <StudentCard key={student.id}  student={student} onDelete={() => setDeleteStudentId(student.id)}/>)}
                    </tbody>
                </table>
                <DeleteStudentModal
                    classUser={{student: deleteStudentId, class: cls}}
                    show={deleteStudentId != -1}
                    onDelete={() => onDelete(deleteStudentId)}
                    close={() => setDeleteStudentId(-1)}/>
            </div>
        }
    </div>)
}