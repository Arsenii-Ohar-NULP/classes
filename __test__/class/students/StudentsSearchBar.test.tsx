import React from 'react';
import {StudentsSearchBar} from "../../../components/class/students/StudentsSearchBar";
import {fireEvent, render, screen, waitFor} from "@testing-library/react";
import {useDispatch} from "react-redux";
    jest.mock('react-redux');

describe('Students Search Bar', () => {
    it('should match a snapshot', () => {
        jest.mocked(useDispatch).mockReturnValue(jest.fn());
        const searchBar = render(<StudentsSearchBar/>)
        expect(searchBar).toMatchSnapshot();
    })

    it('should dispatch `search students`, when field has been changed', async () => {
        const dispatch = jest.fn();
        jest.mocked(useDispatch).mockReturnValue(dispatch);
        render(<StudentsSearchBar/>);
        const input = await screen.findByPlaceholderText('Search for a student');
        fireEvent.change(input, {target: {value: 'Hello'}});

        await waitFor(() => {
            expect(dispatch).toBeCalledTimes(2);
        });
    })
})