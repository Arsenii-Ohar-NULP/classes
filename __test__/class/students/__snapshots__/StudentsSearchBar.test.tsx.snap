// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Students Search Bar should match a snapshot 1`] = `
<div>
  <input
    class="mb-3 form-control"
    placeholder="Search for a student"
    type="text"
  />
</div>
`;
